### DISCUSSION TITLE
Example using Pandas and csv file

### DISCUSSION THREAD

**Discussion by jharming** [NTNU-IHB/PythonFMU#119] [Q&A]
Hi,

I am using windows with anaconda.

I tried to import pandas and then put a csv file into a subfolder called FMU_files.

It does not work like this, so is there an example I can build on ? :)

    from pythonfmu.fmi2slave import Fmi2Slave, Fmi2Causality, Fmi2Variability, Real
    import pandas as pd
    try:
        from sympy import symbols, exp
    except ImportError:  # Trick to be able to generate the FMU without sympy installed
        symbols, exp = None, None

    df_ = pd.read_csv('FMU_files\data.csv', sep=';')

    class Hydraulic(Fmi2Slave):
    """This class is an example to demonstrate installing new Python dependencies.
    The code is not efficient.
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        # self.df = df_
        # self.df['rpm'] = self.df['n']
        self.Q = 1.07
        self.rpm = 1475
        self.H = 32.87
        self.torque = 165.245

        self.register_variable(Real("Q", causality=Fmi2Causality.input))
        self.register_variable(Real("rpm", causality=Fmi2Causality.input))
        self.register_variable(Real("H", causality=Fmi2Causality.output))
        self.register_variable(Real("torque", causality=Fmi2Causality.output))

    def do_step(self, current_time, step_size):
        
        # Q, rpm, H, Torque = symbols("Q, rpm, H, tau")
        #expr = i * a * (1 - exp(-1 * t / tau))
        # self.H = 500. #realOut = expr.evalf(subs={i: self.realIn, a: self.a, t: current_time + step_size, tau: self.tau})
        # self.torque = 300.
        return True

**Comment by markaren** [comment]
Did you embed the FMU_files with the FMU?

`pythonfmu build -f hydraulic.py FMU_files`

**Reply by jharming** [reply]
yes, I did do that. Now I tried to take your example pythonslave.py. 
After each step I build using `pythonfmu build -f pythonslave.py FMU_files` I used FMPy GUI to test the FMU
If I only `import pandas as pd` then I can still build it and run it. When I try to read the csv file I can still build it but not run it. I think it must be the path I am using is wrong. I read using `df = pd.read_csv('FMU_files\hyd_data.csv, sep=';'`)

**Reply by jharming** [reply]
I only read the csv file I dont use it for anything at the moment

**Reply by markaren** [reply]
You probably need a custom getter if you need to fetch data straight from the csv:
`self.register_variable(String("file_content", getter=lambda: data, causality=Fmi2Causality.output))`
where `data` in this case is:
```
with (open('FMU_files/hello.txt', 'r')) as file:
    data = file.read()
```


I don't see why `df = pd.read_csv('FMU_files\hyd_data.csv, sep=';')` should fail. If you look inside the resources folder of the FMU, the file structure should be preserved.

**Reply by markaren** [reply]
Could it be that you are using \ not / ?

**Reply by markaren** [reply]
Actually, forget what I said about the custom getter. The natural approach of reading an csv is to cache data into instance variables.

**Comment by markaren** [comment]
You might also want to look at the CSV builder that ships with pythonfmu. https://github.com/NTNU-IHB/PythonFMU/blob/master/pythonfmu/csvbuilder.py

**Reply by jharming** [reply]
Thanks for the fast replies 👍 
I think I start taking a step back and then I need to understand the syntax a bit better. By the way I took the SympySlave.fmu and checked it at [https://fmu-check.herokuapp.com/](url) here it comes up with an error. I also tried to run it in Simulink but simulink crashed. I can run it just fin using FMPy.

**Reply by markaren** [reply]
> here it comes up with an error

Technically an error with the modelDescription.xml. In practice you probably will not notice. Depends on the importing tool ofcourse.

> I also tried to run it in Simulink but simulink crashed.

Simulink might load some version of Python, which is not compatible. E.g. old version or 32 bit. In which case it will crash. Not much to to about it.

**Comment by markaren** [comment]
So, I think the correct way to read files is to append the value of `self.resources`.
This means that you need to make the call in the context of the slave instance.

```python
class PythonSlaveReadFile(Fmi2Slave):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        with (open(f'{self.resources}/hello.txt', 'r')) as file:
            data = file.read()
```
